Append a self-link to each heading

annevk opened the pull request:
Fixes https://github.com/whatwg/html/issues/46.

domenic commented:
I'm going to try to push a series of commits that get our DOM structure closer to the Bikeshed one. We'll need to update the standard.css file as well I think.

domenic commented:
With the latest commits we have

```html
<h4 id="dom-trees" class="heading"><span class="secno">2.1.3 </span>DOM trees<a href="#dom-trees" class="self-link"></a></h4>
```

which is pretty close to Bikeshed. I didn't create the `<span class="content">` wrapper since it seems unused.

We'd merge this and then transfer some stuff from https://resources.whatwg.org/bikeshed.css to https://resources.whatwg.org/standard.css. I've tested that this works locally. \o/

domenic commented:
The secno fix also helps with https://github.com/whatwg/wattsi/issues/27

annevk commented:
It seems you now overwrite `class="no-num"`. Having `.heading` also seems somewhat redundant, but maybe I'm missing something.

domenic commented:
Ah dang yeah no-num is probably broken.

The class=heading is used by a bunch of the styles in our two stylesheets, but technically we could replace it with a bunch of selectors targeting the individual tag names.

annevk commented:
Okay, I guess we can keep `.heading` for now until someone wants to clean that up... So I guess instead we should get the current attribute value and then append " heading" to it and then set it again?

domenic commented:
I guess we should poke around Wattsi to see if there's some kind of classList abstraction in use, but otherwise that sounds reasonable... Hmm or we could find the place that currently sets it to "no-num" and change it to set to either "heading" or "heading no-num".

annevk commented:
As far as I can tell Wattsi just uses literal comparison on classes. Which is also why changing them can be potentially problematic since things like `class <> 'no-num'` would start yielding true for `heading no-num`...

annevk commented:
It doesn't seem like there's too much class manipulation (in wattsi.pas) so whatever way we go seems doable. Still a little hesitant to add `heading` (when does it apply?) though. Rather not add redundant markup as we already have so much markup.

domenic commented:
Sure, we don't have to add heading; we just have to be careful to update all the CSS.

Do you want to take this over now perchance? :)

annevk commented:
It seems the equivalent of `.heading` is `:has(h2,h3,h4,h5,h6)`, but we'll have to write the latter out as `:has()` is not supported yet.

annevk commented:
I don't really understand why you added `secno` either. It doesn't seem to be needed. Neither of our stylesheets use it  so I'm going to undo that change as well.

annevk commented on src/wattsi.pas:
```diff
@@ -495,6 +495,23 @@    TCrossReferences = record
+                  Scratch.Append('#');
+                  ExtractedData := Element.GetAttribute('id');
+                  Assert(not ExtractedData.IsEmpty);
+                  Scratch.AppendDestructively(ExtractedData); // appending LastSeenHeadingID does not work
```
@Hixie could you explain why simply appending LastSeenHeadingID does not work here? I don't understand how it can be different from GetAttribute('id'). Thanks!

Hixie replied:
What does it do instead?

Hixie replied:
Oh, I see. It's because you're using AppendDestructively, which would destroy LastSeenHeadingID.

Hixie replied:
The reason you can't use LastSeenHeadingID is that it's a local variable, so if you appended it, you would end up having the DOM tree contain a pointer to the stack (rather than the heap). The solution is to not use AsString on the output of EnsureID. Instead, get back a CutRope, which contains pointers to the heap.

Note that EnsureID is itself just going GetAttribute, so it wouldn't make much difference in practice. Also be aware that you can only use AppendDestructively once per reference (hence the name), so if you use LastSeenHeadingID again you'd have to get it back out of the DOM anyway.

annevk commented:
I think this is ready to be merged now, though it would be nice to hear from @Hixie why the obvious code is wrong.

domenic commented:
secno is needed for #27

annevk commented:
Shall we add that as part of fixing that, then? Could even only output it in that mode as that doesn't have single-page. Would save about 700 nodes for normal HTML single-page.

domenic commented:
Sure, works for me. In that case I'll review/probably merge this PR later today.

annevk commented:
Thanks @Hixie, made some changes based on that. I'm not sure I totally follow with regards to stack/heap, but enough to remove a redundant GetAttribute call (although I guess those must be very cheap).

Hixie commented on src/wattsi.pas:
```diff
@@ -637,7 +636,6 @@    TCrossReferences = record
                         begin
                            SecondLI := NewLI.CloneNode(True);
                            ExtractedData := EnsureID(NewLI, 'toc-' + LastSeenHeadingID);
-                           Assert(not ExtractedData.IsEmpty);
```
Why did you remove this assert?

annevk replied:
EnsureID has it, no?

Hixie replied:
The point of the assert here is to make sure EnsureID doesn't break without you noticing. :-)

Hixie commented on src/wattsi.pas:
```diff
@@ -494,7 +494,23 @@    TCrossReferences = record
+
+                  Element.AppendChild(HeadingSelfLink);
+               end;
+
```
i would add a comment here saying ExtractedData is no longer valid past this point

annevk replied:
But it can be assigned again, right? You did not add such comments for other AppendDestructively. Should we comment all of them?

Hixie replied:
It's only non-obvious here because you only destroy in one branch of the if block.

annevk commented on src/wattsi.pas:
```diff
@@ -504,9 +505,7 @@    TCrossReferences = record
                   Scratch := Default(Rope);
                   Scratch.Append('#');
-                  ExtractedData := Element.GetAttribute('id');
-                  Assert(not ExtractedData.IsEmpty);
```
Same as above. EnsureID already asserts.

Hixie replied:
EnsureID is verifying that things within EnsureID are acting as expected.
But here you're outside EnsureID, so you can't trust that EnsureID is correct. Hence the assert.

annevk replied:
So should we add more asserts then? These were the only two places that did that.

Hixie replied:
Add asserts anywhere you have an invariant and you're not confident the rest of the code will keep it, for example when someone breaks the code elsewhere. :-)

annevk commented:
Added the asserts back and added a comment about the validity of ExtractedData. I think this should be good to now.

Corresponding stylesheet PR: https://github.com/whatwg/resources.whatwg.org/pull/21
